dissect/target/filesystems/tar.py

```diff
@@ -1,72 +1,90 @@
+)
 from dissect.target.helpers import fsutil
 
 log = logging.getLogger(__name__)
 
 
 class TarFilesystem(Filesystem):
+    """Filesystem implementation for tar files."""
+
     __fstype__ = "tar"
 
-    def __init__(self, fh, base=None, case_sensitive=True, tarinfo=None, *args, **kwargs):
+    def __init__(
+        self,
+        fh: BinaryIO,
+        base: Optional[str] = None,
+        tarinfo: Optional[tarfile.TarInfo] = None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
         fh.seek(0)
 
         self.tar = tarfile.open(mode="r", fileobj=fh, tarinfo=tarinfo)
-        self.base = base.strip("/") if base else ""
+        self.base = base or ""
 
-        self._fs = VirtualFilesystem(case_sensitive=case_sensitive)
+        self._fs = VirtualFilesystem(alt_separator=self.alt_separator, case_sensitive=self.case_sensitive)
 
         for member in self.tar.getmembers():
-            if member.isdir():
-                continue
-
             mname = member.name.strip("/")
-            if not mname.startswith(self.base):
+            if not mname.startswith(self.base) or mname == ".":
                 continue
 
-            rel_name = mname[len(self.base) :]
+            rel_name = fsutil.normpath(mname[len(self.base) :], alt_separator=self.alt_separator)
 
-            if member.issym():
-                # Do we need to trim the link name for relative tar bases or not?
-                self._fs.symlink(member.linkname, mname)
-            else:
-                file_entry = TarFilesystemEntry(self, rel_name, member)
-                self._fs.map_file_entry(rel_name, file_entry)
-
-        super().__init__(case_sensitive=case_sensitive, *args, **kwargs)
+            entry_cls = TarFilesystemDirectoryEntry if member.isdir() else TarFilesystemEntry
+            file_entry = entry_cls(self, rel_name, member)
+            self._fs.map_file_entry(rel_name, file_entry)
 
     @staticmethod
-    def detect(fh):
+    def detect(fh) -> bool:
         """Detect a tar file on a given file-like object."""
         offset = fh.tell()
         try:
             fh.seek(0)
-            tarfile.TarFile(fileobj=fh, mode="r")
-            return True
+            return tarfile.is_tarfile(fh)
         except Exception as e:
             log.warning("Failed to detect tar filesystem", exc_info=e)
             return False
         finally:
             fh.seek(offset)
 
-    def get(self, path):
+    def get(self, path) -> TarFilesystemEntry:
         """Returns a TarFilesystemEntry object corresponding to the given path."""
         return self._fs.get(path)
 
 
 class TarFilesystemEntry(VirtualFile):
-    def _resolve(self):
+    def _resolve(self) -> TarFilesystemEntry:
         if self.is_symlink():
             return self.readlink_ext()
         return self
 
-    def open(self):
+    def open(self) -> BinaryIO:
         """Returns file handle (file-like object)."""
+        if self.is_dir():
```

Comment:
With the use of `TarFilesystemDirectoryEntry` this directory specific handling here and elsewhere is no longer needed right? As directories in a `TarFilesystem` are either mapped as a `VirtualDirectory` (implicitly) or `TarFilesystemDirectoryEntry` explicitly?

Reply:
It does, since `TarFilesystem` supports symlinks and `self.is_dir()` might dereference to a `TarFilesystemDirectoryEntry`.